src/pwm.rs

```diff
@@ -22,79 +25,70 @@
+///     pwm.try_set_duty(1, max_duty / 4).unwrap();
 /// }
 ///
 /// # use core::convert::Infallible;
-/// # struct KiloHertz(u32);
-/// # trait U32Ext { fn khz(self) -> KiloHertz; }
-/// # impl U32Ext for u32 { fn khz(self) -> KiloHertz { KiloHertz(self) } }
+/// # trait U32Ext { fn khz(self) -> Duration; }
+/// # impl U32Ext for u32 { fn khz(self) -> Duration { Duration::from_nanos( 1_000_000u64 / self as u64 ) } }
 /// # enum Channel { _1, _2 }
 /// # struct Pwm1;
 /// # impl hal::pwm::Pwm for Pwm1 {
+/// #     const CHANNELS: usize = 4;
 /// #     type Error = Infallible;
-/// #     type Channel = Channel;
-/// #     type Time = KiloHertz;
 /// #     type Duty = u16;
-/// #     fn try_disable(&mut self, _: Channel) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_enable(&mut self, _: Channel) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_get_duty(&self, _: Channel) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_disable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_enable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
 /// #     fn try_get_max_duty(&self) -> Result<u16, Self::Error> { Ok(0) }
-/// #     fn try_set_duty(&mut self, _: Channel, _: u16) -> Result<(), Self::Error> { Ok(()) }
-/// #     fn try_get_period(&self) -> Result<KiloHertz, Self::Error> { unimplemented!() }
-/// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<KiloHertz> { Ok(()) }
+/// #     fn try_set_duty(&mut self, _: usize, _: u16) -> Result<(), Self::Error> { Ok(()) }
+/// #     fn try_get_period(&self) -> Result<Duration, Self::Error> { unimplemented!() }
+/// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<Duration> { Ok(()) }
 /// # }
 /// ```
 // unproven reason: pre-singletons API. The `PwmPin` trait seems more useful because it models independent
 // PWM channels. Here a certain number of channels are multiplexed in a single implementer.
 pub trait Pwm {
+    /// Number of available channels
+    const CHANNELS: usize;
+
     /// Enumeration of `Pwm` errors
     type Error;
 
-    /// Enumeration of channels that can be used with this `Pwm` interface
-    ///
-    /// If your `Pwm` interface has no channels you can use the type `()`
-    /// here
-    type Channel;
-
-    /// A time unit that can be converted into a human time unit (e.g. seconds)
-    type Time;
-
     /// Type for the `duty` methods
     ///
     /// The implementer is free to choose a float / percentage representation
     /// (e.g. `0.0 .. 1.0`) or an integer representation (e.g. `0 .. 65535`)
     type Duty;
 
     /// Disables a PWM `channel`
-    fn try_disable(&mut self, channel: Self::Channel) -> Result<(), Self::Error>;
+    fn try_disable(&mut self, channel: usize) -> Result<(), Self::Error>;
```

Comment:
While this is simpler, it does away with having names for channels, which may be more descriptive when reading the code.
Also, this introduces the necessity for a new _runtime_ error: `WrongChannel` (or similar), which is unnecessary when the channel is an `enum` or so.
It would be nicer if this could be enforced at compile time.

Reply:
right, the idea of this is that an _abstract_ `Something<T: Pwm>` can drive N channels using an arbitrary PWM driver, which is somewhere between difficult and impossible using the current system because there's no useful abstraction for the `Channel` associated type bound.